torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -105,41 +105,51 @@ static std::vector<int64_t> seq_to_aten_shape(PyObject *py_seq) {
   return result;
 }
 
-PyObject* tensor_to_numpy(const at::Tensor& tensor) {
+PyObject* tensor_to_numpy(const at::Tensor& tensor, bool force/*=false*/) {
   if (!is_numpy_available()) {
     throw std::runtime_error("Numpy is not available");
   }
-  if (tensor.device().type() != DeviceType::CPU) {
+  
+  bool requires_grad = (at::GradMode::is_enabled() && tensor.requires_grad());
+  if (requires_grad && force == false) {
+    throw std::runtime_error(
+      "Can't call numpy() on Tensor that requires grad. "
+      "Use tensor.detach().numpy() instead.");
+  }
+  auto detached_tensor = (requires_grad && force == true)? tensor.detach() : tensor;  
```

Comment:
You can simplify this code by unconditionally doing `auto new_tensor = tensor.detach().cpu().to_dense()` after all the checks. As these ops will be no-ops if the tensor is already requires_grad=False, on cpu and strided.

Also that will give you a simple way to document what the force keywork is doing.

Reply:
When i implement it like this, i get:

`NotImplementedError: Could not run 'aten::to_dense' with arguments from the 'CPU' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build). If you are a Facebook employee using PyTorch on mobile, please visit https://fburl.com/ptmfixes for possible resolutions. 'aten::to_dense' is only available for these backends: [MkldnnCPU, SparseCPU, SparseCsrCPU, BackendSelect, Named, Conjugate, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, UNKNOWN_TENSOR_TYPE_ID, AutogradMLC, AutogradHPU, AutogradNestedTensor, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, Tracer, UNKNOWN_TENSOR_TYPE_ID, Autocast, Batched, VmapMode].`

Is my build wrong ? Or you just can't call .to_dense() on dense tensor. BTW i expected it to be no-op as well and tried to implement it similarly.

Reply:
Ho, well my bad. Apparently it is not..
Can you just to do it for detach() and cpu() then please?

Also you can link to this issue for the dense part saying that we should update that code once this is solved: https://github.com/pytorch/pytorch/issues/59958